engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala

```diff
@@ -47,55 +46,66 @@ case class ZioExec(runtime: effect.Runtime) extends Exec[ZioExec.IO] {
   /** Execute Zio effect.
     *
     * @param op effect to execute
+    * @param trace object prevents the library from messing up the user's execution trace
     * @return a future containing either a failure or a result
     */
-  override def exec[E, A](op: ZIO[ZEnv, E, A]): Future[Either[E, A]] = {
-    val promise = Promise[Either[E, A]]()
-    runtime.instance.unsafeRunAsync(op) {
-      _.fold(
-        ZioExec.completeFailure(promise, _),
-        ZioExec.completeSuccess(promise, _)
+  override def exec[E, A](
+    op: ZIO[ZAny, E, A]
+  )(implicit trace: Trace): Future[Either[E, A]] =
+    zio.Unsafe.unsafe { implicit unsafe =>
```

Comment:
[zio runtime](https://zio.dev/reference/core/runtime/) says:

> Whenever we run an unsaferun function

Why the heck is that called `unsafe`? What's unsafe on running something, expect it runs!?

Reply:
There is a way to write your programs when you don't execute the effects right away but compose them into a bigger computation. In the zio library, the `ZIO` type represents such a program. The only safe way to run the `ZIO` computation is to return it from the "main" (in zio case it is called `run`) function:

https://github.com/enso-org/enso/blob/6b9cbeacb2caa7606d91fda90ec3ec40acb39c70/lib/scala/project-manager/src/main/scala/org/enso/projectmanager/boot/ProjectManager.scala#L105

It was inspired by pure functional languages like Haskell, where the language simply doesn't have capabilities to perform effects other than returning them from the main function https://hackage.haskell.org/package/base-4.18.0.0/docs/System-IO.html#t:IO and have a similar backdoor https://hackage.haskell.org/package/base-4.18.0.0/docs/System-IO-Unsafe.html

Reply:
I see. I know how actions (IO operations) are done in Haskell. I didn't know about `unsafePerformIO` in Haskell - I do understand why it is called **unsafe** in Haskell.

On the other hand there doesn't seem anything **unsafe** (from JVM perspective) on launching an effect and waiting for its result (if done outside of the effects worker pool). It is certainly not as **unsafe** as using `sun.misc.Unsafe`.

Thanks for the explanation and pointers.